homeassistant/components/switchbee/cover.py

```diff
@@ -0,0 +1,220 @@
+"""Support for SwitchBee cover."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any, TypeVar, Union, cast
+
+from switchbee.api import (
+    SwitchBeeDeviceOfflineError,
+    SwitchBeeError,
+    SwitchBeeTokenError,
+)
+from switchbee.const import SomfyCommand
+from switchbee.device import DeviceType, SwitchBeeShutter, SwitchBeeSomfy
+
+from homeassistant.components.cover import (
+    ATTR_POSITION,
+    CoverDeviceClass,
+    CoverEntity,
+    CoverEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import SwitchBeeCoordinator
+from .entity import SwitchBeeDeviceEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+_DeviceTypeT = TypeVar(
+    "_DeviceTypeT",
+    bound=Union[SwitchBeeShutter, SwitchBeeSomfy],
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up SwitchBee switch."""
+    coordinator: SwitchBeeCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        SwitchBeeCover(device, coordinator)
+        for device in coordinator.data.values()
+        if isinstance(
+            device,
+            (
+                SwitchBeeShutter,
+                SwitchBeeSomfy,
+            ),
+        )
+    )
+
+
+class SwitchBeeCover(SwitchBeeDeviceEntity[_DeviceTypeT], CoverEntity):
```

Comment:
I wonder if it wouldn't be more readable if there were two classes:
- `SwitchBeeCoverEntity` which inherits `SwitchBeeDeviceEntity`
- `SwitchBeeSomfyCoverEntity` which maybe inherits `SwitchBeeEntity`

Let's see how this plays out after you rebase on latest dev.

Reply:
I'll just note that `SwitchBeeSomfy` has no state (its not pulled by the coordinator), it only can be controlled.

Reply:
I still think that instead of having a condition inside every function `if self._device.type == DeviceType.Somfy` it is better to have two classes.

Reply:
I'll try